hazelcast/src/main/java/com/hazelcast/map/impl/record/Records.java

```diff
@@ -50,25 +50,97 @@ public static RecordInfo buildRecordInfo(Record record) {
     }
 
     public static Object getValueOrCachedValue(Record record, SerializationService serializationService) {
-        Object value = record.getCachedValue();
-        if (value == NOT_CACHED) {
-            value = record.getValue();
-        } else if (value == null) {
-            value = record.getValue();
-            if (shouldCache(record, value)) {
-                value = serializationService.toObject(value);
-                record.setCachedValue(value);
+        Object cachedValue = record.getCachedValue();
+        if (cachedValue == NOT_CACHED) {
+            //record does not support caching at all
+            return record.getValue();
+        }
+        for (;;) {
+            if (cachedValue instanceof Thread) {
+                //the cachedValue is either locked by another thread or it contains a wrapped thread
+                cachedValue = ThreadWrapper.unwrapOrNull(cachedValue);
+                if (cachedValue != null) {
+                    //exceptional case: the cachedValue is not locked, it just contains an instance of Thread.
+                    //this can happen when user put an instance of Thread into a map
+                    //(=it should never happen, but never say never...)
+                    return cachedValue;
+                }
+                //it looks like some other thread actually locked the cachedValue
+                Thread.yield();
+            } else if (cachedValue == null) {
+                Object fromCache = tryToLoadIntoCache(record, serializationService);
+                if (fromCache != null) {
+                    return fromCache;
+                }
+            } else {
+                //it's not the 'in-progress' marker && it's not a null -> it has to be actual cachedValue
+                return cachedValue;
             }
+            cachedValue = record.getCachedValue();
         }
-        return value;
     }
 
-    private static boolean shouldCache(Record record, Object value) {
-        boolean isCachableRecordType = record instanceof CachedDataRecordWithStats || record instanceof CachedDataRecord;
-        if (!isCachableRecordType) {
-            return false;
+    private static Object tryToLoadIntoCache(Record record, SerializationService serializationService) {
+        Object valueBeforeCas = record.getValue();
+        if (!shouldCache(valueBeforeCas)) {
```

Comment:
do you think it's alright to merge this and fix the race in a separate issue? 
as the data race is not specific to caching data records.